### PR TITLE
feat: Support config in frontmatter.

### LINKED ISSUE #4630: Diagram specific configuration with YAML (Sankey-beta missing diagram-specific configuration and parsing error)
### Description

**Sankey-beta** diagram parser error when using diagram-specific configuration.

Console output:
```
Parse error on line 1:
%%{init: {'theme': '...
^
Expecting 'SANKEY', got 'NON_ESCAPED_TEXT'
```

### Steps to reproduce

Add diagram-specific configuration:
`%%{init: {'theme': 'default', 'sankey':{ 'width': 400}}}%%`

### Screenshots

With diagram-specific configuration:
image

Without:
image

### Code Sample

```text
%%{init: {'theme': 'default', 'sankey':{ 'width': 400}}}%%
sankey-beta
a 30,b 30,30
b 30,e 64,12
c 70,d 30,30
d 30,e 64,12
c 70,e 64,40
e 64,q 12,12
```

### Setup

- Mermaid version: develop [latest]
- Browser and Version: Chrome 114.0.5735.198

### Suggested Solutions

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by sidharthv96**
## :bookmark_tabs: Summary

Allows defining diagram specific config from the frontmatter. 

```
---
config:
  sankey:
    nodeAlignment: justify
---
sankey-beta
%%data
```   

Resolves #4630 

## :straight_ruler: Design Decisions

The config will be added using the existing `addDirective` function, which has the necessary security checks. 

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [x] :computer: have added necessary unit/e2e tests.
- [x] :notebook: have added documentation. Make sure [`MERMAID_RELEASE_VERSION`](https://github.com/mermaid-js/mermaid/blob/develop/packages/mermaid/src/docs/community/development.md#3-update-documentation) is used for all new features.
- [x] :bookmark: targeted `develop` branch

### CODE REVIEW

**Comment by sidharthv96** [comment]
Related to https://github.com/mermaid-js/mermaid/pull/4748

**Review by aloisklink — CHANGES_REQUESTED** [review]
Two comments:
  - I don't think we should be changing `packages/mermaid/src/config.ts`. Can't we just treat config in the YAML frontmatter the same way we treat `%%{init: ...}` directives?
    That way we can use the already existing [`addDirective()`](https://github.com/mermaid-js/mermaid/blob/72fa3488b5d017362b64f58fccdd49456df50675/packages/mermaid/src/config.ts#L186-L191) function, and we don't need to worry about potential security vulnerabilities in this PR. (we may also need to call `directiveSanitizer()` first).
  - This PR needs documentation on the https://mermaid.js.org/ website (maybe on the https://mermaid.js.org/config/configuration.html page?)

**Comment by sidharthv96** [comment]
> Can't we just treat config in the YAML frontmatter the same way we treat %%{init: ...} directives?

Yes, that was my first approach. But directives are currently not typed properly. So tried this as a POC.

**Comment by aloisklink** [comment]
> Yes, that was my first approach. But directives are currently not typed properly.

Is there any reason why can't just change the `addDirective()` function so that it is typed properly? E.g. something like:

```patch
diff --git a/packages/mermaid/src/config.ts b/packages/mermaid/src/config.ts
index e0066382..a7e6c49b 100644
--- a/packages/mermaid/src/config.ts
+++ b/packages/mermaid/src/config.ts
@@ -8,10 +8,10 @@ export const defaultConfig: MermaidConfig = Object.freeze(config);
 
 let siteConfig: MermaidConfig = assignWithDepth({}, defaultConfig);
 let configFromInitialize: MermaidConfig;
-let directives: any[] = [];
+let directives: MermaidConfig[] = [];
 let currentConfig: MermaidConfig = assignWithDepth({}, defaultConfig);
 
-export const updateCurrentConfig = (siteCfg: MermaidConfig, _directives: any[]) => {
+export const updateCurrentConfig = (siteCfg: MermaidConfig, _directives: MermaidConfig[]) => {
   // start with config being the siteConfig
   let cfg: MermaidConfig = assignWithDepth({}, siteCfg);
   // let sCfg = assignWithDepth(defaultConfig, siteConfigDelta);
@@ -188,7 +188,7 @@ export const sanitize = (options: any) => {
  *
  * @param directive - The directive to push in
  */
-export const addDirective = (directive: any) => {
+export const addDirective = (directive: MermaidConfig) => {
   if (directive.fontFamily) {
     if (!directive.themeVariables) {
       directive.themeVariables = { fontFamily: directive.fontFamily };
```

We may have to add add `// @ts-expect-error` comments, but that's okay, as long as we're improving the existing TypeScript situation (e.g. removing `any` types).

**Comment by sidharthv96** [comment]
> Is there any reason why can't just change the addDirective() function so that it is typed properly?

No reason at all. I guess my brain was in POC mode and didn't want to touch any existing stuff. 

I've cleaned it up now :)

**Review by aloisklink — COMMENTED** [review]
> > No reason at all. I guess my brain was in POC mode and didn't want to touch any existing stuff.
>
> I've cleaned it up now :)

The implementation looks great to me, all it needs is:

- [ ] E2E tests
  - Can you add a test that has both a config directive in YAML frontmatter, and in a `%%{init: }` block, just to confirm it works?
- [ ] an example in the `demos/` folder (or multiple examples), and
- [ ] some documentation in the https://mermaid.js.org site!

If you want to be really fancy, you can rewrite the your commits to split up the `refactor` commits (which have no visible changes to users), `fix:` commits (which fix potential bugs), and the `feat:` commits (which add new features), e.g. something like `refactor: sanitize in addDirective`, `refactor: improve addDirective types`, ... `fix: improve YAML frontmatter type handling`, `feat: support config directives in YAML`, but this is a pretty small PR, so it doesn't matter too much :smile:

FYI, I think this PR will close issue https://github.com/mermaid-js/mermaid/issues/4630! If you agree, feel free to stick it in your PR's description.

File: packages/mermaid/src/assignWithDepth.ts [hunk 1]
```diff
@@ -1,70 +1,24 @@
...
-      } else if (clobber || (typeof dst[key] !== 'object' && typeof src[key] !== 'object')) {
-        dst[key] = src[key];
-      }
-    });
-  }
-  return dst;
-};
+export function assignWithDepth<T>(...data: unknown[]): T {
+  return merge({}, ...data);
+}
```
Comment (sidharthv96) [inline packages/mermaid/src/assignWithDepth.ts:22]: This is something I'm not sure about. 
Previously, assignWithDepth was a security vulnerability when called without sanitizing. 

The only difference is between merging of dissimilar property types (not really applicable for us), and skipping based on depth (which we were not using).

The vulnerabilities did not work with lodash merge.

Comment (sidharthv96) [reply packages/mermaid/src/assignWithDepth.ts]: I'm reverting this change for now. We can fix this in a later PR.

**Review by knsv — APPROVED** [review]

**Comment by sidharthv96** [comment]
Merging myself as @knsv has approved and attempted a merge.

**Pull request merged**